### PR TITLE
Reduce the need for extraneous generics in downstream code.

### PR DESCRIPTION
**Pull request by Pratyush**
## Description

Adds two main APIs:

* `trait FieldWithVar: Field { type Var: FieldVar<Self, Self::BasePrimeField> }`
* `trait CurveWithVar<ConstraintF>: ProjectiveCurve { type Var: CurveVar<Self, ConstraintF> }`

There are refactors to make this work. Primarily: we reimplement everything in terms of 
---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (master)
- [ ] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [ ] Wrote unit tests
- [ ] Updated relevant documentation in the code
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

File: src/fields/nonnative/allocated_field_var.rs [hunk 1]
```diff
@@ -176,6 +176,48 @@ impl<TargetField: PrimeField, BaseField: PrimeField>
...
+                .collect::<Vec<_>>();
+
+            let result = Self {
+                cs,
+                limbs,
+                num_of_additions_over_normal_form,
+                is_in_the_normal_form,
+                target_phantom: PhantomData,
+            };
+            Reducer::<TargetField, BaseField>::post_add_reduce(&mut result)?;
```
Comment (Pratyush) [inline src/fields/nonnative/allocated_field_var.rs:214]: @weikengchen is this okay? I'm using a single reduction to reduce many additions.

Comment (weikengchen) [reply src/fields/nonnative/allocated_field_var.rs:214]: I think it is better to invoke the existing API `add` many times, which, in most cases, also does not introduce constraints.

There are two issues of the change above:
- `num_of_additions_over_normal_form` seems to be reset to a low number. But maybe we should sum together all the `first.num_of_additions_over_normal_form` and get the sum as the new element's `num_of_additions_over_normal_form`. Note that `post_add_reduce` does not always reduce---it only reduces in the extreme cases.
- `post_add_reduce` should better be executed after each addition, in case intermediate additions cause overflow.

Comment (weikengchen) [reply src/fields/nonnative/allocated_field_var.rs:214]: in other words, `post_add_reduce` are cheap.

Comment (Pratyush) [reply src/fields/nonnative/allocated_field_var.rs:214]: > `num_of_additions_over_normal_form` seems to be reset to a low number. But maybe we should sum together all the `first.num_of_additions_over_normal_form` and get the sum as the new element's `num_of_additions_over_normal_form`. Note that `post_add_reduce` does not always reduce---it only reduces in the extreme cases.

Good point, I will sum over the additions too.

> `post_add_reduce` should better be executed after each addition, in case intermediate additions cause overflow.

Is there a formula for that uses `num_of_additions_over_normal_form` to tell me whether a reduction is necessary or not?

Basically I'm trying to avoid generating large `lc`s.

Comment (weikengchen) [reply src/fields/nonnative/allocated_field_var.rs:214]: I got your points. 

Yes, the following code sets a rule. Computing this surfeit every time may be cumbersome, but it may be the more straightforward approach for this PR.  
https://github.com/arkworks-rs/r1cs-std/blob/a2d2d47c0ecd3ea8d2d8f9b68f1d8731abff4c1b/src/fields/nonnative/reduce.rs#L138

Comment (Pratyush) [reply src/fields/nonnative/allocated_field_var.rs]: Ok so in the latest commit I extract that check into a common function, and now invoke it every iteration to see if a reduction is needed or not. If it is, then I invoke reduce, add this intermediate_result to a list of intermediate_results, which I manually add using the `add` operator.

Comment (weikengchen) [reply src/fields/nonnative/allocated_field_var.rs]: I see. The new implementation is reasonable. Basically if `limbs_iter` is too heavy, we sum everything in the `limbs_iter` into a field element that we push to `intermediate_result`, which will be handled later.

**Review by weikengchen — APPROVED** [review]
